[10.0][IMP] account_credit_control : Improve usability

rousseldenis opened the pull request.

rousseldenis committed:
[10.0][IMP] account_credit_control: Improve credit control usability

Introduce 'phone' channel in policies
Add notes on both partner and move lines and display them in credit control lines to help
accountant on day-to-day operations

rousseldenis committed:
[10.0][IMP] account_credit_control: Add move journal on control lines

rousseldenis commented:
@gva-acsone @chrisdec @sbejaoui @sbidoul

pedrobaeza commented:
We did some usability improvement in v11 in #695. You should check it for seeing matching things and align all versions.

rousseldenis commented:
@pedrobaeza Yes, your changes are complementary, so integrate them should not be so complex.

I will note this in my todo for the sprint.

qgroulard commented on account_credit_control/models/account_invoice.py:
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
+                    invoice_move.credit_control_notes
+                break
+
+    @api.model
```
```suggestion
    @api.multi
```

qgroulard commented on account_credit_control/models/account_invoice.py:
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
+    )
+
+    @api.multi
+    def _inverse_credit_control_notes(self):
```
```suggestion:+16
    def _inverse_credit_control_notes(self):
        for invoice in self:
            invoice._get_payable_receivable_move_line().credit_control_notes = invoice.credit_control_notes
```

rousseldenis replied:
Done

qgroulard commented on account_credit_control/models/account_invoice.py:
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
+                break
+
+    @api.multi
+    def _compute_credit_control_notes(self):
```
```suggestion:+11
    def _compute_credit_control_notes(self):
        for invoice in self:
            invoice.credit_control_notes = invoice._get_payable_receivable_move_line().credit_control_notes
```

rousseldenis replied:
Done

qgroulard commented on account_credit_control/models/account_invoice.py:
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
+                break
+
+    @api.model
+    def _get_all_related_move_line(self):
```
No need for this one according to other changes.
Replace it by:
```suggestion:+6
    def_get_payable_receivable_move_line(self):
        self.ensure_one()
        return self.move_id.line_ids.filtered(lambda l: l.account_id.internal_type in ['payable', 'receivable'])
```

qgroulard commented on account_credit_control/models/account_invoice.py:
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
+        ])
+
+    @api.multi
+    def _inverse_credit_control_date(self):
```
```suggestion:+9
    def _inverse_credit_control_date(self):
        for invoice in self:
            invoice._get_payable_receivable_move_line().credit_control_date = invoice.credit_control_date
```

rousseldenis replied:
Done

qgroulard commented on account_credit_control/models/account_invoice.py:
```diff
@@ -30,6 +30,88 @@ class AccountInvoice(models.Model):
+    @api.multi
+    @api.depends('move_id.line_ids.credit_control_date',
+                 'move_id.line_ids.invoice_id')
+    def _compute_credit_control_date(self):
```
```suggestion:+10
    def _compute_credit_control_date(self):
        for invoice in self:
            invoice.credit_control_date = invoice._get_payable_receivable_move_line().credit_control_date
```

qgroulard commented on account_credit_control/models/res_partner.py:
```diff
@@ -51,6 +52,25 @@ class ResPartner(models.Model):
+             " credit control"
+    )
+
+    _sql_constraints = [
```
This sql constraint doesn't make any sense in this module.

rafaelbn commented:
Please @rousseldenis could you rebase? I can test this

rousseldenis commented:
@rafaelbn We plan to work on this on OCA days

rousseldenis commented:
@qgroulard @rafaelbn This is done. Comments attended

qgroulard committed:
[IMP] Several small improvements

Better date format
Don't sum levels when group by
Track status modification
Better breadcrumb
Rename to_be_send -> to_do, sent -> done, date_sent -> date_reminded

houssine78 commented on account_credit_control/models/account_invoice.py:
```diff
@@ -30,6 +30,56 @@ class AccountInvoice(models.Model):
+        self.ensure_one()
+        return self.move_id.line_ids.filtered(
+            lambda l: l.account_id.internal_type in ['payable', 'receivable'])
+
```
from my understanding only receivable accounts are concerned by account_credit_control module... so why also filtering on payable ?

rousseldenis replied:
I'm not quite used to accounting, @chrisdec a little help on this ?

rousseldenis replied:
You're right @houssine78. I'll change it

houssine78 commented on account_credit_control/models/res_partner.py:
```diff
@@ -51,6 +52,20 @@ class ResPartner(models.Model):
+    def _get_default_credit_policy(self):
+        policy_obj = self.env['credit.control.policy']
+        default_policy = policy_obj.search([('default_on_partner', '=', True)])
+        if default_policy:
```
should we take the do_nothing field into account.

rousseldenis replied:
Hi @houssine78, as that field is on the policy, it does not avoid getting a default policy on the partner

houssine78 replied:
@rousseldenis my understanding of this field is to set it at True when the policy become obsolete. as it obsolete does it should be taken as default ? I mean people could forget to unset it on a partner. The idea is why it should continue to be returned as default if it's obsolete

qgroulard replied:
@houssine78 The field 'default_on_partner' is on the policy not on the partner. Thus forgetting to unset it isn't really a concern imo.

houssine78 replied:
ok it's not blocking anyway. ok to move on as it is.

houssine78 commented on account_credit_control/models/credit_control_policy.py:
```diff
@@ -29,6 +29,20 @@ class CreditControlPolicy(models.Model):
+    def _constrains_default_on_partner(self):
+        count = self.search_count([
+            ('default_on_partner', '=', True)
+        ])
```
should we take the do_nothing field into account.

rousseldenis replied:
Same as https://github.com/OCA/account-financial-tools/pull/819/files#r329928573

houssine78 reviewed (approved):
some small remarks but LGTM

qgroulard commented:
@rousseldenis

sbidoul commented on account_credit_control/models/account_move_line.py:
```diff
@@ -0,0 +1,16 @@
+    credit_control_date = fields.Date(
+        string='Credit Control Ignore Before',
+        help="If filled in, specify a pivot date to ignore moves before for "
+             "credit control"
```
This help text is unclear to me.

sbidoul reviewed (commented):
To facilitate review of this relatively big PR, it would be nice to have a functional summary of the changes, possibly with some screenshots.

qgroulard commented:
Here is a list of changes:

- Add credit control notes (free text) on lines, move_lines, invoices and partners;
- Add a credit control date on invoices and move_lines to ignore Credit Control runs until this date;
- Add 'phone' channel;
- Therefore rename to_be_send -> to_do, sent -> done and date_sent -> date_reminded;
- track_visibility="on_change" on credit control line state;
- Display level_lines max level when grouped (groupby in tree view) instead of sum;
- Add a group by move_line_journals on credit control lines;
- Fallback to letter if channel is 'email' and partner has no email;
- Add config on policy to set the policy has default on partners;
- Dates format according to current user settings;
- Fix error message when trying to create a run at a previous date than last run;
- Add specific email for credit control on partner;
